### PR TITLE
chore: remove panic for unimplemented trait dispatch

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Summary\*

This PR replaces a panic with a nicer error plus a link to the relevant issue: https://github.com/noir-lang/noir/issues/4925

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/comptime/interpreter.rs [hunk 1]
```diff
@@ -609,10 +609,13 @@ impl<'a> Interpreter<'a> {
         let rhs = self.evaluate(infix.rhs)?;
 
         // TODO: Need to account for operator overloading
-        assert!(
-            self.interner.get_selected_impl_for_expression(id).is_none(),
-            "Operator overloading is unimplemented in the interpreter"
-        );
+        // See https://github.com/noir-lang/noir/issues/4925
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/comptime/interpreter.rs:612]:
```suggestion
```

File: compiler/noirc_frontend/src/hir/comptime/interpreter.rs [hunk 2]
```diff
@@ -609,10 +609,12 @@ impl<'a> Interpreter<'a> {
         let rhs = self.evaluate(infix.rhs)?;
 
         // TODO: Need to account for operator overloading
-        assert!(
-            self.interner.get_selected_impl_for_expression(id).is_none(),
-            "Operator overloading is unimplemented in the interpreter"
-        );
+        if self.interner.get_selected_impl_for_expression(id).is_some() {
+            return Err(InterpreterError::Unimplemented {
+                item: "Operator overloading is unimplemented in the interpreter. See https://github.com/noir-lang/noir/issues/4925".to_string(),
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/comptime/interpreter.rs:614]:
```suggestion
                item: "Operator overloading in the interpreter".to_string(),
```
The Unimplemented error will display the error as `format!("{item} is currently unimplemented")`

**Review by jfecher — APPROVED** [review]
LGTM

File: compiler/noirc_frontend/src/hir/comptime/interpreter.rs [hunk 3]
```diff
@@ -609,10 +609,12 @@ impl<'a> Interpreter<'a> {
         let rhs = self.evaluate(infix.rhs)?;
 
         // TODO: Need to account for operator overloading
-        assert!(
-            self.interner.get_selected_impl_for_expression(id).is_none(),
-            "Operator overloading is unimplemented in the interpreter"
-        );
+        if self.interner.get_selected_impl_for_expression(id).is_some() {
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/comptime/interpreter.rs:612]:
```suggestion
        // See https://github.com/noir-lang/noir/issues/4925
        if self.interner.get_selected_impl_for_expression(id).is_some() {
```

File: compiler/noirc_frontend/src/hir/comptime/interpreter.rs (hunk 2 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/hir/comptime/interpreter.rs]: I didn't see a good spot to add the issue number - I suppose we could add it as an optional field to the error variant.

File: compiler/noirc_frontend/src/hir/comptime/interpreter.rs (hunk 3 above)
Comment (TomAFrench) [reply compiler/noirc_frontend/src/hir/comptime/interpreter.rs:612]: Re-adding this now it's not in the error message

**Pull request merged**